### PR TITLE
 W tracker

### PR DESCRIPTION
**Pull request by DLWoodruff**
A new extension to output W tracker information.
Should I do a better job isolating the import of pandas?

### CODE REVIEW

File: mpisppy/utils/sputils.py [hunk 1]
```diff
@@ -122,6 +125,97 @@ def reactivate_objs(scenario_instance):
         obj.activate()
 
 
+def id_from_strings(s, vname, idx=None):
```
Comment (jeanpaulwatson) [inline mpisppy/utils/sputils.py:128]: Replace with find_component in Pyomo.

File: mpisppy/utils/sputils.py [hunk 2]
```diff
@@ -122,6 +125,97 @@ def reactivate_objs(scenario_instance):
...
+
+
+def rhos_to_csv(s, filename):
+    """ write the rho values to a csv "fullname", rho
+    Args:
+        s (ConcreteModel): the scenario Pyomo model
+        filenaame (str): file to which to write
+    """
+
+    # This is probably already done somewhere...
```
Comment (jeanpaulwatson) [inline mpisppy/utils/sputils.py:181]: This really should be somewhere else :) in mpi-sppy.

Comment (jeanpaulwatson) [reply mpisppy/utils/sputils.py:181]: And built up on construction.

File: mpisppy/utils/sputils.py [hunk 3]
```diff
@@ -122,6 +125,97 @@ def reactivate_objs(scenario_instance):
...
+                return n
+        raise RuntimeError(f"Cannot find {ndn} in node list for scenario {k}")
+    
+    with open(filename, "w") as f:
+        f.write("fullname,rho\n")
+        for ndn_i, rho in s._mpisppy_model.rho.items():
+            n = _ndn_to_node(ndn_i[0])
+            vdata = n.nonant_vardata_list[ndn_i[1]]
+            fullname = vdata.name
+            f.write(f'"{fullname}",{rho._value}\n')
```
Comment (bknueven) [inline mpisppy/utils/sputils.py:194]: We should use `_mpisppy_data.nonant_indices`: https://github.com/Pyomo/mpi-sppy/blob/main/mpisppy/spbase.py#L292

File: mpisppy/utils/sputils.py [hunk 4]
```diff
@@ -122,6 +125,97 @@ def reactivate_objs(scenario_instance):
...
+    rhodf = pd.read_csv(filename)
+    retlist = list()
+    for idx, row in rhodf.iterrows():
+        fullname = row["fullname"]
+        varname = fullname.split('[')[0]
+        if varname == fullname:
+            varindex = None
+        else:
+            varindex = fullname.split('[')[1].split(']')[0]
+        idval = id_from_strings(s, varname, idx=varindex)
```
Comment (jeanpaulwatson) [inline mpisppy/utils/sputils.py:214]: Use find_component here - no string munging.

File: mpisppy/utils/wtracker.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        self.varnames = [var.name for node in arbitrary_scen._mpisppy_node_list
+                         for (ix, var) in enumerate(node.nonant_vardata_list)]
+
+        
+    def grab_local_Ws(self):
+        """ Get the W values from the PHB that we are following
+            NOTE: we assume this is called only once per iteration
+        """
+        
+        scenario_Ws = {sname: [pyo.value(scenario._mpisppy_model.W[node.name, ix])
```
Comment (jeanpaulwatson) [inline mpisppy/utils/wtracker.py:43]: Could use `_mpisppy_data.nonant_indices` or just `scenario._mpisppy_model.W.values()`

File: mpisppy/utils/wtracker.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        fi = max(1, li - wlen)
+        if li - fi < wlen:
+            return (f"WARNING: Not enough iterations ({cI}) for window len {wlen} and"
+                   f" offsetback {offsetback}\n")
+        else:
+            window_stats = dict()
+            for idx, varname in enumerate(self.varnames):
+                for sname in self.PHB.local_scenario_names:
+                    wlist = [self.local_Ws[i][sname][idx] for i in range(fi, li+1)]
+                    window_stats[(varname, sname)] = [np.mean(wlist), np.std(wlist)]
```
Comment (jeanpaulwatson) [inline mpisppy/utils/wtracker.py:73]: `[np.mean(wlist), np.std(wlist)]` could be a tuple?

**Review by jeanpaulwatson — APPROVED** [review]

File: mpisppy/utils/sputils.py (hunk 3 above)
Comment (DLWoodruff) [reply mpisppy/utils/sputils.py:194]: don.

**Pull request merged**